src/main/java/io/knowledgeassets/myskills/server/communityuser/command/CommunityUserCommandController.java

```diff
@@ -146,7 +146,11 @@ public CommunityUserCommandController(CommunityUserCommandService communityUserC
 			throw new UserCommunityException("The authenticated user must be either a community manager or the one who is leaving the community. " +
 					"And if authenticated user is a community manager she / he cannot remove herself / himself from the community.");
 		} else {
-			communityUserCommandService.remove(communityId, userId);
+			if (securityService.isCommunityManager(communityId)) {
```

Comment:
I've thought some time about on which layer our application should enforce business rules like:

* Community managers are not allowed to remove themselves from their communities.
* Community members are only allowed to remove themselves from communities (they cannot kick others).

I think that controllers are not supposed to handle such business logic because I understood the controller's responsibility to be translating REST API requests to internal service calls and translating internal service results/errors to API response objects. They are just orchestrators of business logic which is encapsulated in the service beans.

Therefore, I'd say that inside this controller method we should only determine the data we need for the internal service calls and move business logic into the service beans. For example, the controller can determine the `Authentication` principal to pass its `userId` to the downstream services. But then those services should implement the logic to determine if, for example, a principal with the user ID 42 is allowed to remove a user with ID 64 from a community with ID 89.

Additionally, I'm not really happy with methods like `isCommunityManager` on a rather generic service like `SecurityService`. If we move community-related business logic to the community service beans - also the decisions regarding community-specific security - we can really encapsulate these implementations in community-specific classes.

What do you think?

Reply:
OK, I think I got your intention now. 😊 Currently, we enforce authorization on the controller layer (e.g. using `PreAuthorize` annotations). You also wanted to keep the more complex community authorization logic on this layer, thus, keeping the service layer clean from any authorization.

With this new knowledge in mind I agree with your implementation. 👍 It's better than scattering authorization across multiple application layers.

I'd just recommend to get rid of the community-specific methods on the `SecurityService`. This service can act as a generic bean which provides access to the current user, etc. But it is not supposed to contain any domain-specific methods. We can create another JIRA story for the refactoring to keep this task out of this feature branch.

This controller method should fetch the `userId` of the `Authentication` principal and then call a method `isCommunityManager(principalUserId, communityId)` on the `CommunityUserQueryService`.

Reply:
OK, I got your point. Now there is already `io.knowledgeassets.myskills.server.security.CurrentUserService` (https://github.com/T-Systems-MMS/skoop-server/blob/master/src/main/java/io/knowledgeassets/myskills/server/security/CurrentUserService.java) to get a current user.

If we want to get rid of `SecurityService` and move domain-specific methods to `CommunityUserQueryService`, `UserPermissionQueryService`, `CommunityQueryService` or other services we will have to establish the dependency between these business services and `CurrentUserService`.

Now I don't see a big profit of such changes. But if it is really worth doing such changes I would prefer to have a separate JIRA story for that.

Reply:
I have thought a bit more about getting rid of `SecurityService`... Now `SecurityService` is the only service which is not used in other services. It is a kind of a service used only by controllers to extend their security abilities with knowledge about business side of the application. So `SecurityService` can be considered to be a middleman between the controller layer and the service layer. Thanks to the `SecurityService` we can say we have all security logic on the controller layer. If we remove `SecurityService` and scatter its methods around business services injecting `CurrentUserService` in them we end up with the security logic present on two layers of the application - the controller layer and the service layer.

What do you think about it?

Reply:
I totally agree that we should not scatter security logic across multiple layers of the application. 👍 

Basically, I don't like the idea of having one service bean which is wired to potentially a lot of other domain-specific services (e.g. community query service).

I guess we agree that we want to keep security logic on the controller layer at the moment, right? - Thus, I think the controllers should make use of several services to collect the data they need to make security decisions. In this example, the procedure in the controller method would be:

* Controller calls a service method to obtain the identity (`User` entity) of the user who performs the REST API request.
* Controller calls another service method to obtain the information if the executing user is a manager in the given community (e.g. calling `CommunityUserQueryService.isCommunityManager(userId, communityId)`)
* Controller decides which service method to call next based upon whether the executing user is a manager or not.

You know what I mean? In this case we have some generic services which provide domain-independent data like the request-executing `User` and some other domain-specific services which can provide domain information, e.g. if a user is a community manager. The controllers bring those two worlds together and make the final authorization decisions.

What do you think? 😃

Reply:
Yeah. You are right. I agree that the security logic is on the controller layer.

I got it at last! 😃 I like the suggested approach. 👍 It should work.

So the controllers will inject `CurrentUserService` and all necessary business-domain services. And the controller is the one who takes over the responsibilities of `SecurityService`.

Could you please create a user story in JIRA for that?

Reply:
By the way, we could also think about injecting the executing user as a controller method argument. I suppose we don't do it at the moment because the [AuthenticationPrincipal](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/core/annotation/AuthenticationPrincipal.html) annotation resolves to the `Jwt` converted from the `Authorization` header. We could provide a custom [HandlerMethodArgumentResolver](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/method/support/HandlerMethodArgumentResolver.html) which can load the real `User` node entity of the authenticated user. - But this is just micro-optimization. 😉 

We can stick with the `CurrentUserService` as a dependency in the controllers at the moment.

Reply:
Alright I'll create a user story for this refactoring and close this one as accepted. 😄

Reply:
Nice idea 👍 I will see what can be done about it. Thanks!